Form helpers elements (#266)
campuscode/rails-guides-pt-BR: pt-BR/form_helpers.md

**NÃO LEIA ESTE ARQUIVO NO GITHUB, OS GUIAS SÃO PUBLICADOS NO https://guiarails.com.br.**
**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**

Auxiliares de formulário de exibição de ação 

Formulários em uma aplicação web é uma inteface essencial para interação do usuário com o sistema. No entanto, pode se tornar entendiante gravar e manter este tipo de marcação devido à necessidade de lidar com a nomeação de controle de formulário e seus inúmeros atributos. O Rails acaba com  essa complexidade pois há um assistente de exibição que gera uma marcação de formulário. Mas para um uso correto, como há casos de uso diferentes, é necessário que os desenvolvedores conheçam as diferenças entre os métodos auxiliares antes de usa-los.

Depois de ler este guia você saberá: 

*Como criar formulários de pesquisa e tipos de semelhantes de formulários genéricos que não representam nenhuma classe específica da sua aplicação. 
*Como criar formulários centrados no modelo pra criar e editar registros específicos no banco de dados.

> *Como gerar caixas de selação de vários tipos de dados.

Reviewer: > Esse eu fiquei bem na dúvida porque é um termo técnico mas ao mesmo tempo não é tanto. Vocês acham que devemos manter ou traduzir `select box` @danielbarroslima e @campuscode/guia-rails-team ?

@HenriqueMorato @danielbarroslima podemos utilizar aquela forma de manter o termo original e também traduzir. Algo como:
```
* Como gerar *select boxes* (caixas de seleção) de vários tipos de dados.
```
O que acham?

> *Como gerar caixas de selação de vários tipos de dados.

Reviewer 2: Acho legal

*Que data e hora os ajudantes do Rails fornecem.
*O que torna um formulário de upload de arquivo diferente.
*Como publicar formulários de recursos externos e especificar a criação de um token de autenticidade (authentic_token).
*Como criar formulários complexos.

> -------------------------------------------------------------------------------------

Reviewer 2: Esse aqui foi?

NOTE: Este guia não pretende ser uma documentação completa dos auxiliares de formulários disponíveis e seus argumentos. Por favor para obter uma  referência completa visite [a documentação da API do Rails](https://api.rubyonrails.org/). 

Trabalhando com formulários basicos.
------------------------------------------------

> O principal auxiliar de formulário é o `form_with` .

Reviewer 2: ```suggestion
O principal auxiliar de formulário (*form helper*) é o `form_with` .
```

```erb
<%= form_with do %>

>   Conteúdo de formulário 

Reviewer 2: ```suggestion
  Conteúdo do formulário 
```

<% end %>
```

> Quando chamado se nenhum argumento para preenchimento como este, é criado uma tag de formulário que, quando enviado, será enviado par página atual. Por exemplo, supondo que a página atual seja a inicial, o HTML gerado teraá a seguinte aparência: 

Reviewer 2: ```suggestion
Quando chamado sem nenhum argumento como este, é criado uma *tag* de formulário que, quando enviado, enviará uma requisição HTTP usando o verbo POST para a página atual. Por exemplo, supondo que a página atual seja a inicial, o HTML gerado terá a seguinte aparência: 
```

Essa frase tinha alguns typos e fiz uma sugestão para manter o POST. Pode nos dizer o que acha?

**NÃO LEIA ESTE ARQUIVO NO GITHUB, OS GUIAS SÃO PUBLICADOS NO https://guiarails.com.br.**
**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**

Action View Form Helpers 
========================

> Formulários em uma aplicação web são uma inteface essencial para interação do usuário com o sistema. No entanto, pode se tornar entendiante manter este tipo de marcação devido à necessidade de lidar com a nomenclatura e controles de formulários e seus inúmeros atributos. O Rails acaba com  essa complexidade pois há um assistente de exibição que gera uma marcação de formulário. Mas para um uso correto, como há casos de uso diferentes, é necessário que os desenvolvedores conheçam as diferenças entre os métodos auxiliares antes de usa-los.

Reviewer: ```suggestion
Formulários em uma aplicação web são uma interface essencial para interação do usuário com o sistema. No entanto, pode-se tornar entendiante manter este tipo de marcação devido à necessidade de lidar com a nomenclatura e controles de formulários e seus inúmeros atributos. O Rails acaba com essa complexidade pois há um assistente de exibição que gera uma marcação de formulário. Mas para um uso correto, como há casos de uso diferentes, é necessário que os desenvolvedores conheçam as diferenças entre os métodos auxiliares antes de usá-los.
```

> Depois de ler este guia você saberá: 

Reviewer: ```suggestion
Depois de ler este guia, você saberá: 
```

*Como criar formulários de pesquisa e tipos de semelhantes de formulários genéricos que não representam *model* específico da sua aplicação. 
*Como criar formulários centrados em *models* pra criar e editar registros específicos no banco de dados.
*Como gerar caixas de selação de vários tipos de dados.
*Que data e hora os ajudantes do Rails fornecem.
*O que torna um formulário de upload de arquivo diferente.
*Como publicar formulários de recursos externos e especificar a criação de um token de autenticidade (authentic_token).
*Como criar formulários complexos.
-------------------------------------------------------------------------------------

NOTE: Este guia não pretende ser uma documentação completa dos auxiliares de formulários disponíveis e seus argumentos. Por favor para obter uma  referência completa visite [a documentação da API do Rails](https://api.rubyonrails.org/). 

> Trabalhando com formulários basicos.

Reviewer: ```suggestion
Trabalhando com formulários básicos.
```

------------------------------------------------

O principal auxiliar de formulário (*form helper*) é o `form_with` .

```erb
<%= form_with do %>
  Conteúdo do formulário 
<% end %>
```

> Quando chamado sem nenhum argumento como este, é criado uma *tag* de formulário que, quando enviado, enviará uma requisição HTTP usando o verbo POST para a  página atual. Por exemplo, supondo que a página atual seja a inicial, o HTML gerado teraá a seguinte aparência: 

Reviewer: ```suggestion
Quando chamado sem nenhum argumento como este, é criado uma *tag* de formulário que, quando enviado, fará uma requisição HTTP usando o verbo POST para a página atual. Por exemplo, supondo que a página atual seja a inicial, o HTML gerado terá a seguinte aparência: 
```
Sugiro trocar o **enviará** por **fará**, pois fica menos redundante. O que acha?

**NÃO LEIA ESTE ARQUIVO NO GITHUB, OS GUIAS SÃO PUBLICADOS NO https://guiarails.com.br.**
**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**

Action View Form Helpers 
========================

> <<<<<<< HEAD

Reviewer: ```suggestion
```

Reviewer 2: > > Efetuei os ajustes solicitados @HenriqueMorato e @tkusuki , obrigado pelas dicas, tinha dado um xabu no pc mas dei um jeito aqui, e continuarei a colaborar \o/ .
> 
> @danielbarroslima dei uma olhada novamente, mas parece que as alterações não foram aplicadas.
> Você chegou a dar o `push` das alterações?
> Pode ter sido algum conflito de mudanças também.
> 
> Pode verificar por aí novamente, por favor?

Parece que um ou outro está faltando mesmo @danielbarroslima você pode dar uma olhada em files changed porque dá pra fazer e ver o merge pelo Github

</form>
```

> Irá notar que o HTML contém um elemento `input` com tipo `hidden`. Esse `input` é importante, isto porqeu o formulário não GET não pode ser enviado com êxito sem ele. O elemento de entrada oculto com o nome `authenticity_token` é um recurso de segurança do  Rails chamado proteção contra falsificação de solicitação entre sites, e os auxiliares de formulário o geram para todos os formulários não GET (desde que esse recurso de segurança esteja ativado). Poderá ler mais sobre isto no guia [segurança em  aplicações Rails](security.html#cross-site-request-forgery-csrf). 

Reviewer 2: ```suggestion
Note que o HTML contém um elemento `input` com tipo `hidden`. Esse `input` é importante, porque o formulário não usa o verbo HTTP de GET então não pode ser enviado com êxito sem ele. Este elemento oculto com o nome `authenticity_token` é um recurso de segurança do Rails chamado **proteção contra falsificação de solicitação entre sites** ([**cross-site request forgery protection**](https://pt.wikipedia.org/wiki/Cross-site_request_forgery)), e os auxiliares de formulário o geram para todos os formulários não GET (desde que esse recurso de segurança esteja ativado). Você pode ler mais sobre isto no guia [segurança em  aplicações Rails](security.html#cross-site-request-forgery-csrf). 
```

Reviewer 2: Fiz uma pequena sugestão e a correção de alguns typos e um espaço a mais. @danielbarroslima sinta-se mais que a vontade para alterar da forma que quiser.

Uma coisa que achei interessante e que talvez possamos fazer em outros locais é incluir o termo da Wikipedia para auxiliar os mais iniciantes, você acha isso legal?

@campuscode/guia-rails-team podem dar uma olhada?

### Formulário de pesquisa genérica 

> Um dos formulários mais básicos que você vê na web é um formulários de pesquisa. Este formulário contém.

Reviewer 2: ```suggestion
Um dos formulários mais básicos que você vê na web é um formulários de pesquisa. Este formulário contém:
```

</form>
```

> Note que o HTML contém um elemento `input` com tipo `hidden`. Esse `input` é importante, porque o formulário não usa o verbo HTTP de então não pode ser enviado com êxito sem ele. Este elemento oculto com o nome `authenticity_token` é um recurso de segurança do  Rails chamado **proteção contra falsificação de solicitação entre sites** ([**cross-site request forgery protection**](https://pt.wikipedia.org/wiki/Cross-site_request_forgery)), e os auxiliares de formulário o geram para todos os formulários não GET (desde que esse recurso de segurança esteja ativado). Poderá ler mais sobre isto no guia [segurança em  aplicações Rails](security.html#cross-site-request-forgery-csrf). 

Reviewer: ```suggestion
Note que o HTML contém um elemento `input` do tipo `hidden`. Este `input` é importante porque o formulário não pode ser enviado com sucesso sem ele, exceto formulários com método GET. Esse elemento oculto com o nome `authenticity_token` é um recurso de segurança do Rails chamado **proteção contra falsificação de solicitação entre sites** ([**cross-site request forgery protection**](https://pt.wikipedia.org/wiki/Cross-site_request_forgery)), e os *helpers* de formulário o geram para todos os formulários não GET (desde que esse recurso de segurança esteja ativado). Você poderá ler mais sobre isso no guia [Segurança em Aplicações Rails](security.html#cross-site-request-forgery-csrf). 
```

### Formulário de pesquisa genérica 

> Um dos formulários mais básicos que você vê na web é um formulários de pesquisa. Este formulário contém:

Reviewer: ```suggestion
Um dos formulários mais básicos que você vê na web é um formulário de pesquisa. Este formulário contém:
```

* Um elemento de formulário com o método GET.

> * Uma etiqueta para entrada.

Reviewer 2: Fiquei na dúvida sobre essa tradução porque acho que podemos manter os termos *label* e *input* por fazer alução a termos técnicos; 

@campuscode/guia-rails-team podem dar uma olhada

> * Um elemento de entrada de texto.

Reviewer 2: Mesma coisa do comentário acima.

</form>
```

> TIP: Passando `url: my_speccified_path` para `form_with` informa ao formulário ode fazer a solicitação. No entanto, conforme explicado abaixo, você também pode passar objetos ActiveRecord para o formulário. 

Reviewer: ```suggestion
TIP: Passando `url: my_speccified_path` para `form_with` indica ao formulário onde fazer a requisição. No entanto, conforme explicado abaixo, você também pode passar objetos ActiveRecord para o formulário. 
```

> TIP: Para cada entrada de formulário, um atributo de ID é gerado a partir de seu nome("q" no exemplo acima). Esses IDs podem ser muito úteis 

Reviewer: ```suggestion
TIP: Para cada entrada de formulário, um atributo ID é gerado a partir de seu nome("q" no exemplo acima). Esses IDs podem ser muito úteis para estilizar CSS ou manipular controles de formulário com JavaScript.
```

> IMPORTANT: Use "GET" como o método para buscas em formulários. Isso permitirá aos usuários marcar uma busca específica e voltar a ela. De forma geral o rails recomenda que você use o verbo correto para a ação desejada.

Reviewer: ```suggestion
IMPORTANT: Use "GET" como o método para buscas em formulários. Isso permitirá aos usuários marcar uma busca específica e depois retornar nessa mesma busca. De forma geral, o Rails recomenda que você utilize o verbo correto para a ação desejada.
```

> ### Auxiliares para gerar elementos de formulário

Reviewer: ```suggestion
### Helpers para gerar elementos de formulário
```

> O Rails fornece  uma série de auxiliares para gerar elementos de formulário, como caixas de seleção, campos de texto e botões de opção. Esses auxilires básicos, com nomes terminados em `_tag`(como `texte_field` e `check_box_tag`), geram apenas um único elemento `<input>`. O primeiro parâmetro para estes será sempre o nome da entrada. Quando o formulário for enviado, o nome será passado junto com os dados do formulário e será direcionado para `params` o controlador com o valor inserido pelo usuário para esse campo. Por exemplo, se o formulário contiver `<%= text_field_tag(:query) %>`, podendo obter o valor deste campo no controlador com `params[:query]` .

Reviewer 2: ```suggestion
O Rails fornece  uma série de auxiliares para gerar elementos de formulário, como caixas de seleção, campos de texto e botões de opção. Esses auxiliares básicos, com nomes terminados em `_tag`(como `text_field` e `check_box_tag`), geram apenas um único elemento `<input>`. O primeiro parâmetro para estes será sempre o nome da entrada. Quando o formulário for enviado, o nome será passado junto com os dados do formulário e será direcionado para `params` o controlador com o valor inserido pelo usuário para esse campo. Por exemplo, se o formulário contiver `<%= text_field_tag(:query) %>`, podendo obter o valor deste campo no controlador com `params[:query]` .
```

Ao nomear entradas, o Rails usa certas convenções que possibilitam enviar parâmetros com valores não escalares, como matrizes ou hashes, que também estarão acessíveis `params`. Poderá ser lido mais sobre eles no capítulo [Noções básicas sobre convenções de nomenclatura de parâmetros](#understanding-parameter-naming-conventions) deste guia. Para detalhes de como usar com precisão esses auxiliares, consulte a [documentação da API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormTagHelper.html).

> #### Caixas de seleção

Reviewer 2: Acho que podemos fazer algo como `Caixas de Seleção (*Checkboxes*)` o que acha @danielbarroslima e @campuscode/guia-rails-team ?

<label for="pet_cat">I own a cat</label>
```

O primeiro parâmetro para `check_box_tag`, é claro, é o nome da entrada. O segundo parâmetro, naturalmente, é o valor da entrada. Este valor será incluído nos dados do formulário (e está presente em `params`) quando a caixa de seleção estiver marcada.

> #### Botões de opção

Reviewer 2: Mesmo caso acima talvez podemos fazer `Botões de Opção (*Radio Buttons*)`

> Os botões de opção embora semelhantes às caixas de seleção, são controles que especificam um conjunto de opções mutuamente exclusivos(ou seja, o usuário pode escolher apenas uma).

Reviewer 2: ```suggestion
Os botões de opção embora semelhantes às caixas de seleção, são controles que especificam um conjunto de opções mutuamente exclusivos (ou seja, o usuário pode escolher apenas uma).
```

<label for="age_adult">I am over 21</label>
```

> Assim como `check_box_tag` o segundo parâmetro para `radio_button_tag` é o valor da entrada. Como esse dois botções compartilham o mesmo nome (`age`), o usuário poderá selecionar apenas um deles e `params[:age]` receberá `"child"` ou `"adult"`.

Reviewer: ```suggestion
Assim como `check_box_tag`, o segundo parâmetro para `radio_button_tag` é o valor da entrada. Como esses dois botões compartilham o mesmo nome (`age`), o usuário poderá selecionar apenas um deles, e `params[:age]` receberá `"child"` ou `"adult"`.
```

> NOTE: Sempre use etiquetas para a caixa de seleção e botões de opção. Eles associam o texto a uma opção específica e, ao expandir a região clicável, facilita o clique dos usuários nas entradas.

Reviewer: ```suggestion
NOTE: Sempre use *labels* para a caixa de seleção e botões de opção. Eles associam o texto a uma opção específica e, ao expandir a região clicável, facilita o clique dos usuários nas entradas.
```

<input id="product_discount" max="100" min="1" name="product[discount]" type="range" />
```

> Entradas ocultas não são mostradas ao usuário, mas retêm dados como qualquer entrada de texto. Os valores dentro deles podem ser alterados com JavaScript.

Reviewer: ```suggestion
Entradas ocultas não são exibidas ao usuário, mas retêm dados como qualquer entrada de texto. Os valores dentro deles podem ser alterados com JavaScript.
```

> IMPORTANT: As entradas de pesquisa, telefone, data, hora, cor, data e hora local e data, mês, semana, URL, email, número e intervalo são controles HTML5. Se você precisar que sua aplicação tenha uma experiência consistente em navegadores antigos, precisará de um polyfill HTML5(fornecido por css e/ou JavaScript). Definiticamente, [não faltam soluções disponíveis](https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills), embora uma ferramenta popular no momento seja o [Modernizr](https://modernizr.com/), que fornece uma maneira simples de adicionar funcionalidade com base na presença de recursos HTML5 detectados.

Reviewer: ```suggestion
IMPORTANT: As entradas de pesquisa, telefone, data, hora, cor, data e hora local e data, mês, semana, URL, email, número e intervalo são controles HTML5. Se você precisar que sua aplicação tenha uma experiência consistente em navegadores antigos, precisará de um polyfill HTML5 (fornecido por css e/ou JavaScript). Definiticamente, [não faltam soluções disponíveis](https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills), embora uma ferramenta popular no momento seja o [Modernizr](https://modernizr.com/), que fornece uma maneira simples de adicionar funcionalidade com base na presença de recursos HTML5 detectados.
```

> TIP: Se você estiver usando campos de entradas de senha (para qualquer finalidade), comvêm configurar sua aplicação para impedir que esses parâmetros sejam registrados. Você pode aprender sobre este assunto no guia [Protegendo aplicações Rails](security.html#logging)

Reviewer 2: ```suggestion
TIP: Se você estiver usando campos de entradas de senha (para qualquer finalidade), convêm configurar sua aplicação para impedir que esses parâmetros sejam registrados. Você pode aprender sobre este assunto no guia [Protegendo aplicações Rails](security.html#logging)
```
